### PR TITLE
Support CallerID as query rule condition

### PR DESCRIPTION
**Pull request by adsr**
Signed-off-by: Adam Saponara <as@php.net>

### CODE REVIEW

File: go/vt/vttablet/tabletserver/rules/rules_test.go [hunk 1]
```diff
@@ -642,12 +656,14 @@ var invalidjsons = []InvalidJSONCase{
...
 	{`[{"Description": 1 }]`, "want string for Description"},
 	{`[{"RequestIP": 1 }]`, "want string for RequestIP"},
+	{`[{"CallerID": 1 }]`, "want string for CallerID"},
 	{`[{"User": 1 }]`, "want string for User"},
 	{`[{"Query": 1 }]`, "want string for Query"},
 	{`[{"Plans": 1 }]`, "want list for Plans"},
 	{`[{"TableNames": 1 }]`, "want list for TableNames"},
 	{`[{"BindVarConds": 1 }]`, "want list for BindVarConds"},
 	{`[{"RequestIP": "[" }]`, "could not set IP condition: ["},
+	{`[{"CallerID": "[" }]`, "could not set CallerIDf condition: ["},
```
Comment (deepthi) [inline go/vt/vttablet/tabletserver/rules/rules_test.go]:
```suggestion
	{`[{"CallerID": "[" }]`, "could not set CallerID condition: ["},
```

**Comment by adsr** [comment]
Hi @deepthi This is useful for us to create read-only db users. I believe it is possible to do this via table ACLs but it is cumbersome to make thousands of entries. With this patch it can be achieved in a single query rule entry.

**Comment by dweitzman** [comment]
Table ACLs can operate on groups, incidentally. In our table acls we just grant read access to members of the "readers" group, and our auth plugin for vtgate ensures that all users have immediate caller IDs that include the appropriate groups for that user.

**Comment by adsr** [comment]
@dweitzman Is there a way to apply an ACL to all tables without having to maintain a list of tables?

**Comment by dweitzman** [comment]
Yes. I think "%" is the wildcard that matches all tables in a table acl
config

**Comment by sougou** [comment]
Blacklist rules are mainly meant for preventing things, like stopping toxic queries. Hopefully, the '%' approach will work for table ACLs. Even if it doesn't, I'd recommend extending that functionality instead.

Having said that, I won't be opposed to adding this to the rules functionality, but for the use case where one wants to block all queries coming from a user gone rogue.

In general, both rules and ACLs need work. They use config files, and are not cloud-friendly. We need to move these into the topo, and add workflows to maintain them.

**Comment by dweitzman** [comment]
> In general, both rules and ACLs need work. They use config files, and are not cloud-friendly. We need to move these into the topo, and add workflows to maintain them.

Not to get too far off topic, but for security-conscious folks, public companies subject to Sarbanes-Oxley, CI/CD nerds, etc., it's probably good idea to be very deliberate about separating machine-written information (like tablet registration) from things that require distributed consensus (like topo information that needs to be locked during shard splits or other operations) from things that are authored by humans and simply need to be eventually consistent (compiled binaries, ACLs, blacklists, etc.)

People who want to have a clean control / auditing / rollout / rollback story for eventually-consistent human-written logic and configuration might prefer that those deploy as signed archives using standard rollout mechanisms rather than pulling from a more dynamic database like etcd or zk.

**Comment by sougou** [comment]
In youtube, these were deployed through topo (chubby). However, the configs only contained role names. Actual membership to those roles was maintained by a separately audited system.

**Comment by dweitzman** [comment]
> In youtube, these were deployed through topo (chubby). However, the configs only contained role names. Actual membership to those roles was maintained by a separately audited system.

If I can change the list of roles in chubby, couldn't I widen the ACL scopes to include all of engineering for all tables? Seems like the source of ACL config in chubby would need to be very carefully protected and audited (to the exact same standard as the separate systems for checking role membership, probably).

**Comment by sougou** [comment]
Right. One doesn't simply write into Chubby. There are workflows that controlled what went into those ACLs. It was part of the schema deployment process, which was audited through code reviews. This was the reason these parts are not open-sourced because they're google-specific.

**Comment by adsr** [comment]
Thanks, I overlooked the wildcard feature in table ACLs. That works for us. I'm OK to close this PR in that case.

**Comment by sougou** [comment]
Closing this for now. We can revisit if a new use case arises.

**Pull request closed without merging**